lib/Dialect/SV/Transforms/HWSpecialize.cpp

```diff
@@ -0,0 +1,298 @@
+  typeConverter.addConversion([=](hw::ArrayType type) {
+    return evaluateParametricType(loc, parameters, type).getValue();
+  });
+
+  // Valid target types.
+  typeConverter.addConversion([](mlir::IntegerType type) { return type; });
+}
+
+// Specializes the provided 'base' module into the 'target' module. By doing
+// so, we create a new module which
+// 1. has no parameters
+// 2. has a name composing the name of 'base' as well as the 'parameters'
+// parameters.
+// 3. Has a top-level interface with any parametric types resolved.
+// 4. Any references to module parameters have been replaced with the
+// parameter value.
+static LogicalResult specializeModule(OpBuilder builder, ArrayAttr parameters,
+                                      HWModuleOp source, HWModuleOp &target) {
+  auto *ctx = builder.getContext();
+  // Update the types of the source module ports based on evaluating any
+  // parametric in/output ports.
+  auto ports = source.getPorts();
+  for (auto &in : llvm::enumerate(source.getFunctionType().getInputs())) {
+    auto resType =
+        evaluateParametricType(source.getLoc(), parameters, in.value());
+    if (failed(resType))
+      return failure();
+    ports.inputs[in.index()].type = resType.getValue();
+  }
+  for (auto &out : llvm::enumerate(source.getFunctionType().getResults())) {
+    auto resType =
+        evaluateParametricType(source.getLoc(), parameters, out.value());
+    if (failed(resType))
+      return failure();
+    ports.outputs[out.index()].type = resType.getValue();
+  }
+
+  // Create the specialized module using the evaluated port info.
+  target = builder.create<HWModuleOp>(
+      source.getLoc(),
+      StringAttr::get(ctx, generateModuleName(source, parameters)), ports);
+
+  // Erase the default created hw.output op - we'll copy the correct operation
+  // during body elaboration.
+  (*target.getOps<hw::OutputOp>().begin()).erase();
+
+  // Clone body of the source into the target.
+  BlockAndValueMapping mapper;
+  for (auto &&[src, dst] :
+       llvm::zip(source.getArguments(), target.getArguments()))
+    mapper.map(src, dst);
+  builder.setInsertionPointToStart(target.getBodyBlock());
+  for (auto &op : source.getOps())
+    builder.clone(op, mapper);
+
+  // We've now created a separate copy of the source module with a rewritten
+  // top-level interface. Next, we enter the module to convert parametric types
+  // within operations.
+  RewritePatternSet patterns(ctx);
+  TypeConverter t;
+  populateTypeConversion(target.getLoc(), t, parameters);
+  patterns.add<EliminateParamValueOpPattern>(ctx, parameters);
+  patterns.add<NarrowArrayGetIndexPattern>(ctx);
+  patterns.add<ParametricTypeConversionPattern>(ctx, t, parameters);
+  ConversionTarget convTarget(*ctx);
+  convTarget.addLegalOp<hw::HWModuleOp>();
+  convTarget.addIllegalOp<hw::ParamValueOp>();
+
+  // Generic legalization of converted operations.
+  convTarget.addDynamicallyLegalDialect<comb::CombDialect, hw::HWDialect,
```

Comment:
If you don't want to limit this to ops in these dialects, you can use:
`target.markUnknownOpDynamicallyLegal([](Operation *op) { return !isParametricOp(op); });`

(I only recently figured this out myself)

Reply:
Nice, this was **exactly** what i needed here - i found it ridiculous that dialects had to be explicitly named... kind of defeats the point of the whole thing! thank you!